File: backend/app/tools/scanner.py
import asyncio
from typing import Any, Dict, List
from urllib.parse import urlparse


COMMON_PORTS = [80, 443, 8080, 3000, 5000]


async def _probe_port(hostname: str, port: int, timeout: float = 1.5) -> Dict[str, Any]:
    try:
        connection = asyncio.open_connection(hostname, port)
        reader, writer = await asyncio.wait_for(connection, timeout=timeout)
        writer.close()
        await writer.wait_closed()
        return {"port": port, "state": "open"}
    except Exception:
        return {"port": port, "state": "closed"}


async def port_scan(target_url: str) -> Dict[str, Any]:
    parsed = urlparse(target_url if "://" in target_url else f"https://{target_url}")
    hostname = parsed.hostname or ""

    if not hostname:
        return {"host": "", "ports": [], "open_ports": [], "error": "Port scan could not start because the hostname was invalid."}

    try:
        results = await asyncio.gather(*[_probe_port(hostname, port) for port in COMMON_PORTS])
    except Exception as error:
        return {
            "host": hostname,
            "ports": [],
            "open_ports": [],
            "error": f"Port scan failed for {hostname}: {error}",
        }

    open_ports = [result["port"] for result in results if result["state"] == "open"]
    return {"host": hostname, "ports": results, "open_ports": open_ports}


def format_port_logs(scan_result: Dict[str, Any]) -> List[str]:
    if scan_result.get("error"):
        return [f"[EXECUTE] PORT_SCAN: {scan_result['error'].upper()}."]

    host = scan_result.get("host", "").upper() or "UNKNOWN HOST"
    open_ports = scan_result.get("open_ports", [])
    if open_ports:
        return [
            f"[EXECUTE] PORT_SCAN: {host} RESPONDED ON {', '.join(str(port) for port in open_ports)}.",
            f"[EXECUTE] PORT_SCAN: CLOSED OR FILTERED PORTS REMAIN UNDER OBSERVATION.",
        ]

    return [
        f"[EXECUTE] PORT_SCAN: NO COMMON WEB PORTS RESPONDED ON {host}.",
        "[EXECUTE] PORT_SCAN: TARGET MAY BE FILTERED, IDLE, OR FRONTED BY UPSTREAM CONTROLS.",
    ]

File: aether/src/lib/scanner/tools/scanner.ts
import { createHash } from "crypto";
import type { PortScanResult } from "../types";

const COMMON_PORTS = [80, 443, 8080, 3000, 5000];

function probePort(hostname: string, port: number, timeout = 1500): Promise<{ port: number; state: string }> {
  return new Promise((resolve) => {
    const net = require("net");
    const socket = new net.Socket();
    const timer = setTimeout(() => {
      socket.destroy();
      resolve({ port, state: "closed" });
    }, timeout);

    socket.connect(port, hostname, () => {
      clearTimeout(timer);
      socket.destroy();
      resolve({ port, state: "open" });
    });

    socket.on("error", () => {
      clearTimeout(timer);
      resolve({ port, state: "closed" });
    });
  });
}

export async function portScan(targetUrl: string): Promise<PortScanResult> {
  let hostname = "";
  try {
    const parsed = new URL(targetUrl.includes("://") ? targetUrl : `https://${targetUrl}`);
    hostname = parsed.hostname;
  } catch {
    return { host: "", ports: [], open_ports: [], error: "Invalid URL" };
  }

  if (!hostname) return { host: "", ports: [], open_ports: [], error: "Invalid hostname" };

  try {
    const results = await Promise.all(COMMON_PORTS.map((p) => probePort(hostname, p)));
    return {
      host: hostname,
      ports: results,
      open_ports: results.filter((r) => r.state === "open").map((r) => r.port),
    };
  } catch (error) {
    return { host: hostname, ports: [], open_ports: [], error: `Port scan failed: ${error}` };
  }
}

export function formatPortLogs(result: PortScanResult): string[] {
  if (result.error) return [`[EXECUTE] PORT_SCAN: ${result.error.toUpperCase()}.`];
  const host = result.host.toUpperCase() || "UNKNOWN HOST";
  if (result.open_ports.length > 0) {
    return [
      `[EXECUTE] PORT_SCAN: ${host} RESPONDED ON ${result.open_ports.join(", ")}.`,
      "[EXECUTE] PORT_SCAN: CLOSED OR FILTERED PORTS REMAIN UNDER OBSERVATION.",
    ];
  }
  return [
    `[EXECUTE] PORT_SCAN: NO COMMON WEB PORTS RESPONDED ON ${host}.`,
    "[EXECUTE] PORT_SCAN: TARGET MAY BE FILTERED, IDLE, OR FRONTED BY UPSTREAM CONTROLS.",
  ];
}
